yarn-project/simulator/src/public/enqueued_calls_processor.ts

```diff
@@ -161,8 +171,49 @@ export class EnqueuedCallsProcessor {
     let isFromPrivate = true;
     let returnValues: NestedProcessReturnValues[] = [];
     let revertReason: SimulationError | undefined;
+
+    const nonRevertibleNullifiersFromPrivate = publicKernelOutput.endNonRevertibleData.nullifiers
+      .filter(n => !n.isEmpty())
+      .map(n => n.value);
+    const _revertibleNullifiersFromPrivate = publicKernelOutput.end.nullifiers
+      .filter(n => !n.isEmpty())
+      .map(n => n.value);
+
+    // During SETUP, non revertible side effects from private are our "previous data"
+    const prevAccumulatedData = publicKernelOutput.endNonRevertibleData;
+    const previousValidationRequestArrayLengths = PublicValidationRequestArrayLengths.new(
+      publicKernelOutput.validationRequests,
+    );
+
+    const previousAccumulatedDataArrayLengths = PublicAccumulatedDataArrayLengths.new(prevAccumulatedData);
+    const innerCallTrace = new PublicSideEffectTrace();
+    const enqueuedCallTrace = new PublicEnqueuedCallSideEffectTrace(
+      /*startSideEffectCounter=*/ 0,
+      previousValidationRequestArrayLengths,
+      previousAccumulatedDataArrayLengths,
+    );
+    const trace = new DualSideEffectTrace(innerCallTrace, enqueuedCallTrace);
+
+    // Transaction level state manager that will be forked for revertible phases.
+    const txStateManager = AvmPersistableStateManager.newWithPendingSiloedNullifiers(
+      this.worldStateDB,
+      trace,
+      nonRevertibleNullifiersFromPrivate,
+    );
+    // TODO(dbanks12): insert all non-revertible side effects from private here.
+
     for (let i = 0; i < phases.length; i++) {
```

Comment:
I wonder if eventually (when the code gets simpler) it would make sense to just unroll the phases and have

// SETUP
...

// APP LOGIC
...

// TEARDOWN

it would be in principle much more readable (if each part is small) instead of having a loop with special cases in the body.